scipy/optimize/_qap.py

```diff
@@ -0,0 +1,408 @@
+            P = J * init_weight + (1 - init_weight) * K
+        else:
+            P = J
+        const_sum = A21 @ B21.T + A12.T @ B12
+        grad_P = np.inf  # gradient of P
+        n_iter = 0  # number of FW iterations
+
+        # OPTIMIZATION WHILE LOOP BEGINS
+        while grad_P > eps and n_iter < maxiter:
+            # computing the gradient of f(P) = -tr(APB^tP^t)
+            delta_f = (const_sum + A22 @ P @ B22.T + A22.T @ P @ B22)
+            # run hungarian algorithm on gradient(f(P))
+            rows, cols = linear_sum_assignment(obj_func_scalar * delta_f)
+            Q = np.zeros((n_unseed, n_unseed))
+            Q[rows, cols] = 1  # initialize search direction matrix Q
+
+            def f(x):  # computing the original optimization function
+                return obj_func_scalar * (
+                    np.trace(A11.T @ B11)
+                    + np.trace(np.transpose(x * P + (1 - x) * Q) @ A21 @ B21.T)
+                    + np.trace(np.transpose(x * P + (1 - x) * Q) @ A12.T @ B12)
+                    + np.trace(
+                        A22.T
+                        @ (x * P + (1 - x) * Q)
+                        @ B22
+                        @ np.transpose(x * P + (1 - x) * Q)
+                    )
+                )
+
+            alpha = minimize_scalar(
+                f, bounds=(0, 1), method="bounded"
+            ).x  # computing the step size
+            P_i1 = alpha * P + (1 - alpha) * Q  # Update P
+            grad_P = np.linalg.norm(P - P_i1)
+            P = P_i1
+            n_iter += 1
+        # end of FW optimization loop
+
+        row, col = linear_sum_assignment(
+            -P
+        )  # Project onto the set of permutation matrices
+        perm_inds_new = np.concatenate(
+            (np.arange(n_seeds), np.array([x + n_seeds for x in col]))
+        ).astype(int)
+
+        score_new = np.trace(
+            np.transpose(cost_matrix)
+            @ dist_matrix[np.ix_(perm_inds_new, perm_inds_new)]
+        )  # computing objective function value
+
+        if obj_func_scalar * score_new < obj_func_scalar * score:  # minimizing
+            score = score_new
+            perm_inds = np.zeros(n, dtype=int)
+            perm_inds[permutation_cost] = permutation_dist[perm_inds_new]
+
+    permutation_cost_inv = np.argsort(permutation_cost)
+    cost_matrix = cost_matrix[
+        np.ix_(permutation_cost_inv, permutation_cost_inv)
+    ]
+    permutation_dist_inv = np.argsort(permutation_dist)
+    dist_matrix = dist_matrix[
+        np.ix_(permutation_dist_inv, permutation_dist_inv)
+    ]
+
+    score = np.trace(
+        np.transpose(cost_matrix) @ dist_matrix[np.ix_(perm_inds, perm_inds)]
+    )
+
+    res = {"col_ind": perm_inds, "score": score, "nit": n_iter}
+
+    return OptimizeResult(res)
+
+
+def _check_init_input(init, n):
+    row_sum = np.round(np.sum(init, axis=0), decimals=2)
+    col_sum = np.round(np.sum(init, axis=1), decimals=2)
```

Comment:
Consider comparing the difference between the sum and a 1 against a tolerance rather than rounding and checking equality.

More importantly, be sure to use `atleast_2d` and check for too many dimensions.